docker/archive/archive_src.go

```diff
@@ -0,0 +1,79 @@
+package archive
+
+import (
+	"fmt"
+	"io"
+	"os"
+
+	"github.com/containers/image/manifest"
+	"github.com/containers/image/types"
+)
+
+const temporaryDirectoryForBigFiles = "/var/tmp" // Do not use the system default of os.TempDir(), usually /tmp, because with systemd it could be a tmpfs.
+
+type archiveImageSource struct {
+	ref     archiveReference
+	img     *Image
+	tarFile io.Closer
+}
+
+func newImageSource(ctx *types.SystemContext, ref archiveReference) (types.ImageSource, error) {
+	tarFile, err := os.Open(ref.path)
+	if err != nil {
+		return nil, err
+	}
+
+	img, err := NewImageSource(tarFile)
+	if err != nil {
+		return nil, err
+	}
+
+	return &archiveImageSource{
+		ref:     ref,
+		img:     img,
+		tarFile: tarFile,
+	}, nil
+}
+
+// Reference returns the reference used to set up this destination.  Note that this should directly correspond to user's intent,
+// e.g. it should use the public hostname instead of the result of resolving CNAMEs or following redirects.
+func (s *archiveImageSource) Reference() types.ImageReference {
+	return s.ref
+}
+
+// Close removes resources associated with an initialized ImageSource, if any.
+func (s *archiveImageSource) Close() {
+	s.tarFile.Close()
+}
+
+// GetManifest returns the image's manifest along with its MIME type. The empty string is returned if the MIME type is unknown.
+// It may use a remote (= slow) service.
+func (s *archiveImageSource) GetManifest() ([]byte, string, error) {
+	// XXX: We aren't caching this because GetManifest() specifically states
+	// that it can use a slow service. Caching makes this uglier (and also
+	// makes implementing read-write images harder).
+	m, err := s.img.GenerateManifest()
+	if err != nil {
+		return nil, "", err
+	}
+
+	// FIXME: Do we want to suppport manifest lists?
```

Comment:
I believe it does support manifest lists. By construction the format **is** a manifest list, but it only contains one entry (there's some support in the daemon IIRC because people have created multi-arch images).